src/Raven.Server/Documents/Indexes/IndexStorage.cs

```diff
@@ -706,8 +706,11 @@ private void WriteReferencesForSingleCollectionInternal(Tree referencesTree, str
                     var key = keys.Key;
                     foreach (var referenceKey in keys.Value)
                     {
-                        collectionTree.MultiAdd(referenceKey, key);
-                        referencesTree.MultiAdd(key, referenceKey);
+                        if (collectionTree.MultiExists(referenceKey, key) == false)
+                            collectionTree.MultiAdd(referenceKey, key);
```

Comment:
As I understand `MultiAdd(referenceKey, key)` was effectively no-op in the tested reindexing scenario, right?

This case is supposed to be handled here:
https://github.com/ravendb/ravendb/blob/e60b383cccf6820e1bde91ef387de326f1af8650/src/Voron/Data/BTrees/Tree.MultiTree.cs#L82-L88

Isn't it checking the same as `MultiExists()` you added? The difference that is `MultiAdd()` does it in one call. 

Why `MultiExists / MultiAdd` is faster than just  `MultiAdd` which does the existence of the same value anyway?

Reply:
This isn't the issue.
The issue is that `MultiAdd` is **always** modifying the page which forces us to write the page.

https://github.com/ravendb/ravendb/blob/fc318e40649881a130450611c98d1417a5eaf019/src/Voron/Data/BTrees/Tree.MultiTree.cs#L48

https://github.com/ravendb/ravendb/blob/fc318e40649881a130450611c98d1417a5eaf019/src/Voron/Data/BTrees/Tree.MultiTree.cs#L59

Reply:
Shouldn't we then try to fix `MultiAdd()` then? To avoid doing any modifications if this is a no op?

Please also note than in 6.2 the tree state is handled differently, and we have different code:

https://github.com/ravendb/ravendb/blob/e60b383cccf6820e1bde91ef387de326f1af8650/src/Voron/Data/BTrees/Tree.MultiTree.cs#L46-L50

CC @redknightlois

Reply:
Also worth reconsidering if this change should go to 5.4

Reply:
> Shouldn't we then try to fix `MultiAdd()` then? To avoid doing any modifications if this is a no op?

This should anohter method because this is only needed when handing references.

Reply:
FWIW - fixing `MultiAdd` is the better option, but it complicates things significantly for the common case where we are always modifying.